openingd/dualopend.c

```diff
@@ -95,8 +123,1303 @@ struct state {
+
+	return serial_id % 2 == 0;
+}
+
+static size_t psbt_input_weight(struct wally_psbt *psbt,
+				size_t in)
+{
+	size_t weight;
+	u16 max_witness_len;
+	bool ok;
+
+	/* txid + txout + sequence */
+	weight = (32 + 4 + 4) * 4;
+	weight +=
+		(psbt->inputs[in].redeem_script_len +
+			(varint_t) varint_size(psbt->inputs[in].redeem_script_len)) * 4;
+
+	ok = psbt_input_get_max_witness_len(&psbt->inputs[in],
+					    &max_witness_len);
+	assert(ok);
+	weight += max_witness_len;
+	return weight;
+}
+
+static bool check_balances(struct wally_psbt *psbt,
+			   bool check_opener_balance,
+			   bool do_full_tx_check,
+			   u32 feerate_per_kw_funding)
+{
+	struct amount_sat side_input_amt, side_output_amt,
+			  input_amt, output_amt, fee, diff;
+	size_t weight, input_count = 0, output_count = 0;
+	bool ok;
+
+	if (check_opener_balance)
+		weight = common_weight(psbt->num_inputs, psbt->num_outputs);
+	else
+		weight = 0;
+
+	/* Find the total input and output sums for this participant */
+	input_amt = AMOUNT_SAT(0);
+	side_input_amt = AMOUNT_SAT(0);
+	for (size_t i = 0; i < psbt->num_inputs; i++) {
+		struct amount_sat amt = psbt_input_get_amount(psbt, i);
+
+		/* Add to total balance check */
+		ok = amount_sat_add(&input_amt, input_amt, amt);
+		assert(ok);
+		if (is_openers(&psbt->inputs[i].unknowns) == check_opener_balance) {
+			ok = amount_sat_add(&side_input_amt, side_input_amt, amt);
+			assert(ok);
+
+			weight += psbt_input_weight(psbt, i);
+			input_count++;
+		}
+	}
+	output_amt = AMOUNT_SAT(0);
+	side_output_amt = AMOUNT_SAT(0);
+	for (size_t i = 0; i < psbt->num_outputs; i++) {
+		struct amount_sat amt = psbt_output_get_amount(psbt, i);
+		/* Add to total balance check */
+		ok = amount_sat_add(&output_amt, output_amt, amt);
+		assert(ok);
+		if (is_openers(&psbt->outputs[i].unknowns) == check_opener_balance) {
+			ok = amount_sat_add(&side_output_amt, side_output_amt, amt);
+			assert(ok);
+
+			weight += (8 + psbt->tx->outputs[i].script_len +
+					varint_size(psbt->tx->outputs[i].script_len)) * 4;
+			output_count++;
+		}
+	}
+
+	/* Inputs must exceed outputs */
+	if (do_full_tx_check && !amount_sat_greater(input_amt, output_amt))
+		return false;
+
+	/* No inputs, no outputs. Nothing to see here */
+	if (input_count + output_count == 0)
+		return true;
+
+	/* Find difference, or fail if too small */
+	if (!amount_sat_sub(&diff, side_input_amt, side_output_amt))
+		return false;
+
+	/* Figure out the fee for their side */
+	fee = amount_tx_fee(feerate_per_kw_funding, weight);
+
+	/* Can they cover their fee? */
+	return amount_sat_greater_eq(diff, fee);
+}
+
+
+/*~ This is the key function that checks that their configuration is reasonable:
+ * it applied for both the case where they're trying to open a channel, and when
+ * they've accepted our open. */
+static bool check_configs(struct state *state,
+			  const struct channel_config *remoteconf,
+			  bool am_opener)
+{
+	struct amount_sat capacity;
```

Comment:
This function recently changed in openingd (anchor outputs).  Might need to extract that and share?